presto-tests/src/test/java/com/facebook/presto/resourcemanager/TestDistributedClusterStatsResource.java

```diff
@@ -79,25 +111,77 @@ private String getResourceFilePath(String fileName)
+    @Test(timeOut = 220_000)
     public void testClusterStatsRedirectToResourceManager()
             throws Exception
     {
-        runToExecuting(client, coordinator, "SELECT avg(totalprice) from tpch.sf100.orders");
-
-        // Sleep to allow query to make some progress
-        sleep(SECONDS.toMillis(20));
-
-        ClusterStatsResource.ClusterStats clusterStats = getClusterStats(true);
-
+        waitUntilCoordinatorsDiscoveredHealthyInRM(SECONDS.toMillis(15));
+        runToFirstResult(client, coordinator2, "SELECT * from tpch.sf102.orders");
+        runToFirstResult(client, coordinator2, "SELECT * from tpch.sf100.orders");
+        runToFirstResult(client, coordinator2, "SELECT * from tpch.sf101.orders");
+        waitForGlobalQueryViewInCoordinator(3, RUNNING, coordinator1, SECONDS.toMillis(20));
+        runToQueued(client, coordinator2, "SELECT * from tpch.sf100.orders");
+        waitForGlobalQueryViewInCoordinator(1, QUEUED, coordinator1, SECONDS.toMillis(20));
+        ClusterStatsResource.ClusterStats clusterStats = getClusterStats(true, coordinator1);
         assertNotNull(clusterStats);
         assertTrue(clusterStats.getActiveWorkers() > 0);
-        assertEquals(clusterStats.getRunningQueries(), 1);
+        assertEquals(clusterStats.getRunningQueries(), 3);
         assertEquals(clusterStats.getBlockedQueries(), 0);
-        assertEquals(clusterStats.getQueuedQueries(), 0);
+        assertEquals(clusterStats.getQueuedQueries(), 1);
+        assertEquals(clusterStats.getAdjustedQueueSize(), 0);
+    }
+
+    private void waitForGlobalQueryViewInCoordinator(int numberOfRunningQueries, QueryState state, TestingPrestoServer coordinator, long timeoutInMillis)
+            throws InterruptedException, TimeoutException
+    {
+        long deadline = System.currentTimeMillis() + timeoutInMillis;
+        int globalQueryCount = 0;
+        while (System.currentTimeMillis() < deadline) {
+            Map<ResourceGroupId, ResourceGroupRuntimeInfo> resourceGroupRuntimeInfoSnapshot = coordinator.getResourceGroupManager().get().getResourceGroupRuntimeInfosSnapshot();
+            ResourceGroupRuntimeInfo resourceGroupRuntimeInfo = resourceGroupRuntimeInfoSnapshot.get(new ResourceGroupId(RESOURCE_GROUP_GLOBAL));
+            if (resourceGroupRuntimeInfo == null) {
+                continue;
+            }
+            globalQueryCount = getQueryCount(state, resourceGroupRuntimeInfo);
+            if (globalQueryCount == numberOfRunningQueries) {
+                return;
+            }
+            sleep(100);
+        }
+        throw new TimeoutException(format("Global Query Count: %s after %s ms", globalQueryCount, timeoutInMillis));
+    }
+
+    private int getQueryCount(QueryState state, ResourceGroupRuntimeInfo resourceGroupRuntimeInfo)
+    {
+        int queryCount = 0;
+        switch (state) {
+            case RUNNING:
+                queryCount = resourceGroupRuntimeInfo.getDescendantRunningQueries();
```

Comment:
why descendant running queries and not just running queries? I think this should return the running queries/queued queries as per the state and the globalQueryCount to aggregated it.

Reply:
We are using getDescendantRunningQueries here to consider the  sum of all  running queries under that resource group.  In the test we are using this to get all running queries under "global"

Reply:
This will be good if we are passing a specific resourceGroupRuntimInfo (i.e. for global), but this will result in wrong result if we end up having the last resourceGroupRuntimeInfo being some intermediate group (i.e. global.user-user). So better we use the running queries and queued queries here and aggregate the values where we use the function.

Reply:
Ah I see your point now. The method name is not clarifying the intent here. I have refactored the code a bit. Please check again.